File: crates/draco-cpp-test-bridge/cpp/decode_drc.cpp
// Decodes a .drc file N times and nothing else -- the C++ half of the
// callgrind comparison, matching examples/decode_drc.rs step for step.
//
// Built directly against a Draco checkout rather than through the Rust
// bridge, because the comparison runs where valgrind does (Linux/WSL) and the
// bridge links a Windows library. See PERFORMANCE.md, the callgrind round,
// for the build line and why both sides must decode the same file.
//
//   g++ -O2 -DNDEBUG -std=c++17 -I<src> -I<build> decode_drc.cpp \
//       -L<build> -ldraco -o decode_drc
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <vector>

#include "draco/compression/decode.h"

int main(int argc, char** argv) {
    if (argc < 2) {
        std::fprintf(stderr, "usage: decode_drc <file.drc> [iters]\n");
        return 2;
    }
    const int iters = argc > 2 ? std::atoi(argv[2]) : 1;

    std::ifstream file(argv[1], std::ios::binary);
    if (!file) {
        std::fprintf(stderr, "cannot open %s\n", argv[1]);
        return 2;
    }
    std::vector<char> encoded((std::istreambuf_iterator<char>(file)),
                              std::istreambuf_iterator<char>());

    int64_t points = 0, faces = 0;
    for (int i = 0; i < iters; ++i) {
        draco::DecoderBuffer buffer;
        buffer.Init(encoded.data(), encoded.size());
        draco::Decoder decoder;
        auto result = decoder.DecodeMeshFromBuffer(&buffer);
        if (!result.ok()) {
            std::fprintf(stderr, "decode failed: %s\n",
                         result.status().error_msg());
            return 1;
        }
        points = result.value()->num_points();
        faces = result.value()->num_faces();
    }
    std::fprintf(stderr, "cpp decoded %s: %lld points / %lld faces x%d\n",
                 argv[1], (long long)points, (long long)faces, iters);
    return 0;
}

File: crates/draco-cpp-test-bridge/examples/decode_drc.rs
//! Decodes a `.drc` file N times and nothing else.
//!
//! The counterpart of `cpp/decode_drc.cpp`, and the Rust half of the
//! callgrind comparison: one decode of one file, no encoder, no harness, no
//! C++ bridge -- so it builds anywhere `draco-core` does, including the WSL
//! side where `valgrind` lives and the Windows-built Draco library does not.
//!
//! Under callgrind use `1` iteration: the tool counts instructions, so
//! repetition buys nothing and costs minutes.
//!
//! ```text
//! cargo run --release --example decode_drc -- grid_s5.drc 1
//! valgrind --tool=callgrind --callgrind-out-file=rust.out ./decode_drc grid_s5.drc 1
//! ```
use draco_core::decoder_buffer::DecoderBuffer;
use draco_core::mesh::Mesh;
use draco_core::mesh_decoder::MeshDecoder;

fn main() {
    let mut args = std::env::args().skip(1);
    let path = args.next().expect(".drc path");
    let iters: u32 = args.next().and_then(|s| s.parse().ok()).unwrap_or(1);

    let encoded = std::fs::read(&path).expect("read .drc");
    let mut shape = (0, 0);
    for _ in 0..iters {
        let mut buffer = DecoderBuffer::new(&encoded);
        let mut mesh = Mesh::new();
        let mut decoder = MeshDecoder::new();
        decoder
            .decode(&mut buffer, &mut mesh)
            .expect("decode failed");
        shape = (mesh.num_points(), mesh.num_faces());
    }
    eprintln!(
        "rust decoded {path}: {} points / {} faces x{iters}",
        shape.0, shape.1
    );
}
